Crane/app/src/main/java/androidx/compose/samples/crane/details/DetailsActivity.kt

```diff
@@ -177,58 +173,69 @@ fun DetailsContent(
 }
 
 @Composable
-private fun CityMapView(latitude: String, longitude: String) {
-    // The MapView lifecycle is handled by this composable. As the MapView also needs to be updated
-    // with input from Compose UI, those updates are encapsulated into the MapViewContainer
-    // composable. In this way, when an update to the MapView happens, this composable won't
-    // recompose and the MapView won't need to be recreated.
-    val mapView = rememberMapViewWithLifecycle()
-    MapViewContainer(mapView, latitude, longitude)
-}
-
-@Composable
-private fun MapViewContainer(
-    map: MapView,
+fun CityMapView(
     latitude: String,
-    longitude: String
+    longitude: String,
+    cameraState: CameraPositionState = rememberSaveable(saver = CameraPositionState.Saver) { CameraPositionState() },
+    onMapLoaded: () -> Unit = {}
 ) {
-    val cameraPosition = remember(latitude, longitude) {
+    val cityLocation = remember(latitude, longitude) {
         LatLng(latitude.toDouble(), longitude.toDouble())
     }
 
-    LaunchedEffect(map) {
-        val googleMap = map.awaitMap()
-        googleMap.addMarker { position(cameraPosition) }
-        googleMap.moveCamera(CameraUpdateFactory.newLatLng(cameraPosition))
+    cameraState.position = CameraPosition.fromLatLngZoom(
```

Comment:
This will be executed on every recomposition. Is it really needed or can go in an effect? Regardless, maybe add a comment explaining it? Thanks

IIUC, this will reset a zoom that the user could've set to `InitialZoom`, and we might not want that, right?

Reply:
So it only recomposes this latitude or longitude changes but I agree it looks a bit fragile. 

Ideally we would just not expose the cameraPositionState and use the built in initialiser for position, the problem is then we can't test that the map is being positioned correctly. I really wanted this Composable to kind be as described "show a map at this position with a marker" but in order to achieve the "at this position" part and be testable it needs to initialise the camera somewhere.

LaunchedEffect doesn't feel right because then the setting of the position is delayed
DisposableEffect doesn't feel right because there is nothing to dispose
SideEffect runs too late so has similar results to LaunchedEffect
I have done a trick with remember here to make sure it only reinitialises when cityLocation changes but it feels like a hack.

The other thing that could be done is we just pass in an initialised camerastate in both tests and code but that feels like it won't be testing much.